llvm/lib/CASObjectFormats/NestedV1.cpp

```diff
@@ -748,138 +779,75 @@ static Error decomposeAndSortEdges(
   if (!Block.edges_size())
     return Error::success();
 
-  // Collect edges and targets, filtering out duplicate symbols.
-  SmallVector<const jitlink::Edge *, 16> Edges;
-  Edges.reserve(Block.edges_size());
-  struct TargetAndSymbol {
-    TargetRef Target;
-    const jitlink::Symbol *Symbol;
-  };
   struct EdgeTarget {
-    Optional<TargetRef> Target; // None if this is an abstract backedge.
-    jitlink::Edge::AddendT Addend = ~0U;
+    const jitlink::Symbol *Symbol;
+    // Index in the edge array for the edge that this target is part of.
+    size_t EdgeIdx;
   };
-  bool HasAbstractBackedge = false;
-  SmallDenseMap<cas::CASID, size_t, 16> SymbolIndex;
-  SmallDenseMap<const jitlink::Edge *, EdgeTarget, 16> EdgeTargets;
-  SmallVector<TargetAndSymbol, 16> TargetData;
-  for (const jitlink::Edge &E : Block.edges()) {
-    Edges.push_back(&E);
-    const jitlink::Symbol &S = E.getTarget();
+  SmallVector<EdgeTarget, 16> EdgeTargets;
+  Fixups.reserve(Block.edges_size());
+  TIs.reserve(Block.edges_size());
+
+  visitEdgesInStableOrder(Block, [&](const jitlink::Edge &E) {
+    Fixups.push_back(Fixup{E.getKind(), E.getOffset()});
+    // We'll fill the real index later, using '~0U' to mark as 'unset'.
+    TIs.push_back(TargetInfo{E.getAddend(), ~0U});
```

Comment:
Can you use a name for the `~0U` both here and below?

e.g.,
```
constexpr unsigned AbstractBackedgeIndexPlaceholder = ~0U`;
```

Up to you what to name it... or if you really prefer the magic constant in both places I can live with that.

Reply:
Good idea!